packages/ember-application/lib/system/engine-instance.js

```diff
@@ -0,0 +1,79 @@
+/**
+@module ember
+@submodule ember-application
+*/
+
+import EmberObject from 'ember-runtime/system/object';
+import Registry from 'container/registry';
+import ContainerProxy from 'ember-runtime/mixins/container_proxy';
+import RegistryProxy from 'ember-runtime/mixins/registry_proxy';
+import { get } from 'ember-metal/property_get';
+import { set } from 'ember-metal/property_set';
+import run from 'ember-metal/run_loop';
+
+/**
+  The `EngineInstance` encapsulates all of the stateful aspects of a
+  running `Engine`.
+
+  @public
+  @class Ember.EngineInstance
+  @extends Ember.Object
+  @uses RegistryProxyMixin
+  @uses ContainerProxyMixin
+*/
+
+let EngineInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+  /**
+    The base `Engine` for which this is an instance.
+
+    @property {Ember.Engine} engine
+    @private
+  */
+  base: null,
+
+  init() {
+    this._super(...arguments);
+
+    let base = get(this, 'base');
+
+    if (!base) {
+      base = get(this, 'application');
+      set(this, 'base', base);
```

Comment:
I agree this is an anti-pattern that I'd like to deprecate. See this discussion with @mixonic - https://github.com/emberjs/ember.js/pull/12685#discussion-diff-46772712

I was previously setting `base` in `ApplicationInstance` before calling `_super`, but I moved away from that to be ES6 Class-compliant. Not sure which is the larger concern.

I had been setting `base` This previously